cosmos/web/views.py

```diff
@@ -57,23 +58,124 @@ def workflow(name):
+        last_url = url_for("cosmos.query", workflow_name=workflow_name, stage_name=stage_name,
+                           old_page=max_page, old_keyword=keyword, old_in_page=in_page,
+                           sorting=sorting, order=order) if page != max_page else None
+
+        # this will change only the url for the column currently used for sorting
+        order_cycle = {None: "asc", "asc": "desc", "desc": None}
+        ordering_for_urls = {colname: order_cycle[order] if good == sorting else "asc"
+                             for colname, good in zip(colnames, names_internal)}
+        ordering_urls = {colname: url_for(f"cosmos.stage",
+                                          workflow_name=workflow_name, stage_name=stage_name, in_page=in_page,
+                                          page=page, keyword=keyword, sorting=good, order=ordering_for_urls[colname]
+                                          ) if good else None
+                         for colname, good in zip(colnames, names_internal)}
+
+        submitted = [t for t in tasks_paginated if t.status == TaskStatus.submitted]
+        submitted = tasks_paginated
         jm = JobManager(get_submit_args=None, logger=None)
 
         f = attrgetter("drm")
         drm_statuses = {}
         for drm, tasks in it.groupby(sorted(submitted, key=f), f):
             drm_statuses.update(jm.get_drm(drm).drm_statuses(list(tasks)))
 
-        return render_template("cosmos/stage.html", stage=stage, drm_statuses=drm_statuses)
+        url_query = url_for('cosmos.query', old_page=page, old_keyword=keyword, sorting=sorting, order=order,
+                            workflow_name=workflow_name, stage_name=stage_name, old_in_page=in_page)
+
+        return render_template("cosmos/stage.html", stage=stage, drm_statuses=drm_statuses, in_page=in_page,
+                               tasks_on_page=tasks_paginated, max_page=max_page, colnames=colnames,
+                               ordering_urls=ordering_urls, page=page, url_query=url_query,
+                               first_url=first_url, prev_url=prev_url, next_url=next_url, last_url=last_url,
+                               workflow_name=workflow_name, stage_name=stage_name, keyword=keyword)
         # x=filter(lambda t: t.status == TaskStatus.submitted, stage.tasks))
 
+    @bprint.route("/workflow/<workflow_name>/<stage_name>/query", methods=["GET", "POST"])
+    def query(workflow_name, stage_name):
+        page = request.args.get("old_page", 1, int)
+        keyword = request.args.get("old_keyword", "", type=str)
+        in_page = request.args.get("old_in_page", 40, type=int)
+        order = request.args.get("order", 1, int)
+        sorting = request.args.get("old_sorting", None, type=str)
+
+        if request.form.get("submit_page") == "Go to page":
+            page = request.form.get("page")
+
+        if request.form.get("submit_search") == "Search":
```

Comment:
it seems this should be
if
elif
elif..
else:
  raise AssertionError('invalid form')